### PR TITLE
Use const instead of macros

### PR DESCRIPTION
**Pull request by sosthene-nitrokey**
This is pretty much a complete rewrite. It gives a much nicer API to
understand from the documentation. It also means that it doesn't rely on
\#[doc(hidden)] methods (except for `Interchange::const_new` so that it
can be removed once the required compiler features are stabilized)

Because everything doesn't need to be `'static` anymore, it can also be
tested efficiently with [Loom](https://lib.rs/crates/loom), ensuring that there are no possible race
conditions.

It is a significantly breaking change

There may be some possible improvement. For example the `claims` mechanism could be implemented in the `Channel` struct and use RAII to be made safe and remove the need for `reset-claims`. This could make the API 100% safe which would be nice, but it would not be zero cost, given that it's unlikely a `Requester` or a `Responder` would ever be dropped in most use cases.

### CODE REVIEW

**Comment by sosthene-nitrokey** [comment]
I've updated the PR with [this trick](https://old.reddit.com/r/rust/comments/wwsddr/const_fn_parameter_usage_propagation_throws_errors/ilndzec/) which makes it possible to have a `const` constructor and completely remove all macros.

I also added an RAII mechanism instead of the `reset_claims` which should not be that big of an overhead. It also makes `Channel::split` safe which means that no public API is unsafe, and that `Channel` can be used by itself without using the `Interchange` structure.

**Comment by sosthene-nitrokey** [comment]
I've made a branch that also removes the `Clone` trait bound on the request and response structs. https://github.com/sosthene-nitrokey/interchange/pull/1

**Comment by robin-nitrokey** [comment]
It would be helpful to also implement `Default` for `Interchange` and `Channel`.

**Comment by sosthene-nitrokey** [comment]
Isn't it already the case?

- https://github.com/trussed-dev/interchange/pull/5/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R405
- https://github.com/trussed-dev/interchange/pull/5/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R913

**Comment by robin-nitrokey** [comment]
Ah, sorry, I probably looked at the wrong branch.

File: src/lib.rs [hunk 1]
```diff
@@ -481,67 +755,163 @@ impl<I: Interchange> Responder<I> {
...
-            Err(())
+            Err(Error)
         }
     }
 
     /// Respond to a request.
     ///
     /// If efficiency is a concern, or responses need multiple steps to
-    /// construct, use `response_mut` and `send_response.
+    /// construct, use `with_response_mut` or `response_mut` and `send_response`.
```
Comment (robin-nitrokey) [inline src/lib.rs:771]: s/response_mut/response/

File: src/lib.rs [hunk 2]
```diff
@@ -481,67 +755,163 @@ impl<I: Interchange> Responder<I> {
...
+                    }
+                })
             }
+            Ok(unsafe { self.data_mut().rp_mut() })
+        } else {
+            Err(Error)
+        }
+    }
+
+    /// Send a response that was already placed in the channel using `response_mut` or
```
Comment (robin-nitrokey) [inline src/lib.rs:836]: s/response_mut/response/

Comment (sosthene-nitrokey) [reply src/lib.rs:836]: I've renamed the method to `response_mut` for consistency with the requester

File: src/lib.rs (hunk 1 above)
Comment (sosthene-nitrokey) [reply src/lib.rs:771]: I've renamed the method to `response_mut` for consistency with the requester

**Comment by nickray** [comment]
Not worried about the overhead of the atomic bools to implement the RAII, just to state the obvious though, it's not guaranteed destructors (drop) are run. Seems fine for the intended use cases.

Overall, I think these are good changes, particularly adding the testing with Loom. Do we have any chance of upstreaming constness as needed to Loom, or is this a property of the implementations there?

Also.. did Loom catch any bugs already?

File: src/lib.rs [hunk 3]
```diff
@@ -481,67 +761,163 @@ impl<I: Interchange> Responder<I> {
...
+/// #
+/// static interchange: Interchange<Request, Response,10> = Interchange::new();
+///
+/// for i in 0..10 {
+///     let rq: Requester<'_, Request, Response>;
+///     let rp: Responder<'_, Request, Response>;
+///     (rq, rp) = interchange.claim().unwrap() ;
+/// }
+/// ```
+pub struct Interchange<Q, A, const N: usize> {
```
Comment (nickray) [inline src/lib.rs]: I presume Q = query/question, and A = answer? Maybe use longer names in the struct definition (and switch to the short ones in the implementation)? In any case, please add a comment on what the names mean, so it's easier to understand.

**Comment by sosthene-nitrokey** [comment]
[This](https://github.com/tokio-rs/loom/blob/master/src/rt/atomic.rs#L213) is the implementation of `new` for  loom's atomic mocks.

It registers itself to loom's runtime so I doubt it will ever be made `const`.
It *may* be possible to make it `const` by making the registration happen lazily on the first operation, but I don't know if it's worth it or if it can work.

File: src/lib.rs (hunk 3 above)
Comment (sosthene-nitrokey) [reply src/lib.rs]: You're right, I'm not very satisfied with Q,A since everything else mentions Request/Response. I'll replace it with `Rq` and `Rp`

**Comment by sosthene-nitrokey** [comment]
I also added an `InterchangeRef` type that removes the need to have a `const` parameter everywhere.

```rust
static INTERCHANGE_INNER: Interchange<Request, Response, 1> = Interchange::new();

// The size of the interchange is absent from the type
static INTERCHANGE: InterchangeRef<'static, Request, Response> = INTERCHANGE_INNER.as_interchange_ref();

let (mut rq, mut rp) = INTERCHANGE.claim().unwrap();
```

**Pull request merged**